### ISSUE TITLE
Triggers including unicode characters does not work

### ISSUE THREAD

**Issue by marjoweb** [federico-terzi/espanso#1347]
**The following match does not work:**
  # Print the yesterday's date
  - trigger: ":igår"
    replace: "{{mydate}}"
    vars:
      - name: mydate
        type: date
        params:
          format: "%Y-%m-%d"
          offset: -86400

When changing trigger to "igar" or "tomorrow" everything works as expected.
Is there a workaround for this? I tried using a regex trigger, but could not get that to work either. Perhaps I'm using the wrong syntax.

Any help is as always much appreciated! :)

**Comment by mturilin** [comment]
I have the same issue

**Comment by Aldekein** [comment]
I am also blocked by this issue, since I was planning a migration to espanso, but I use a lot of Ukrainian shortcuts.

**Comment by Karkir72** [comment]
This also makes it difficult for me to work. I would like to see support for other languages for shortcuts sooner.

**Comment by federico-terzi** [comment]
Hey folks!

Thanks for raising this bug, it's a tricky one to solve as the behavior is really platform-dependant. Which operating systems are you using?

**Comment by assarj** [comment]
> Hey folks!
> 
> Thanks for raising this bug, it's a tricky one to solve as the behavior is really platform-dependant. Which operating systems are you using?

MacOS Monterey 12.6.5

**Comment by Karkir72** [comment]
> Hey folks!
> 
> Thanks for raising this bug, it's a tricky one to solve as the behavior is really platform-dependant. Which operating systems are you using?

I am very glad that you responded! thank you! I am using windows 10

**Comment by fluxzero** [comment]
I have the same issue here using Korean triggers and Linux (EndeavourOS with KDE Plasma 5.27.5).

**Comment by grigoryosifov** [comment]
The issue still persists on macOS Sonoma 14.1. Any updates?

**Comment by smeech** [comment]
I suspect this is related to the other issues with non-English keyboards.
https://github.com/espanso/espanso/issues/345
https://github.com/espanso/espanso/issues/498
https://github.com/espanso/espanso/issues/511
https://github.com/espanso/espanso/issues/996
https://github.com/espanso/espanso/issues/1112
https://github.com/espanso/espanso/issues/1133
https://github.com/espanso/espanso/issues/1286

? https://github.com/espanso/espanso/issues/628
? https://github.com/espanso/espanso/issues/1028
? https://github.com/espanso/espanso/issues/1327

**Comment by netpastor** [comment]
HI! Are there any updates for Russian keyboard and letters?

**Comment by luftaquila** [comment]
Same issue here for the Korean triggers(MacOS).

**Comment by mthies-unibi** [comment]
I can confirm this problem on macOS Sonoma 14.4.1 with a German keyboard layout, where all non-ASCII characters (like the German umlauts Ä, Ö, Ü, ä, ö, ü, ß) are **not** recognized in triggers. This affects normal triggers, literal parts of regex triggers, and even capturing groups in regex triggers.

For macOS this is caused by a bug in line 77 of [native.mm](https://github.com/espanso/espanso/blob/91bdd9666de6c0f45024bfb8037324882e57c3af/espanso-detect/src/mac/native.mm#L77).

```
const char *chars = [event.characters UTF8String];
strncpy(inputEvent.buffer, chars, 23);
inputEvent.buffer_len = event.characters.length;
```

This uses the number of **Unicode characters** produced by the keyboard event, but the Rust code expects the number of **bytes** after UTF-8 encoding the characters. Only for ASCII characters, which are encoded as a single byte in UTF-8, there is no difference.

Changing the last line above to

```
inputEvent.buffer[23] = 0;  // ensure NUL at end, even if strncpy had to truncate
inputEvent.buffer_len = strlen(inputEvent.buffer);  // measure length in bytes, instead of UTF-8 chars
```

fixes the problem for me. Triggers with Unicode characters work for the German umlauts with their dedicated keys in my keyboard layout and also for other international characters that require the Option key to be held down.

Note that this does **not** fix any problems with dead keys in triggers. If I type a dead <kbd>^</kbd> followed by <kbd>a</kbd> the trigger detection code in Espanso misses the `^` completely and only sees the character `a`. Line 75 of [native.mm](https://github.com/espanso/espanso/blob/91bdd9666de6c0f45024bfb8037324882e57c3af/espanso-detect/src/mac/native.mm#L75) could be fixed by using `[event charactersByApplyingModifiers: event.modifierFlags]` so that the trigger detection code sees the dead key as a separate character `^` immediately (even before another character has been typed), but then the backspacing over the trigger would remove one character too many in most cases.

**Comment by rpnfan** [comment]
Same problem on Windows 11. Needs a fix IMO :-)

**Comment by blza** [comment]
> Changing the last line above to
> 
> ```
> inputEvent.buffer[23] = 0;  // ensure NUL at end, even if strncpy had to truncate
> inputEvent.buffer_len = strlen(inputEvent.buffer);  // measure length in bytes, instead of UTF-8 chars
> ```

These lines were added in commit 8daadcc9 and merged into master. So now building latest version (2.2.2 as of now) according to Compilation.md fixes the issue for me on Sonoma.

**Comment by grischard** [comment]
I still can't use `trigger: "fée"` or any of the accented typo autocorrections in, for example, https://hub.espanso.org/typofixer-pt-br . This is espanso 2.3.0 on macOS 15.7.2, using dead keys to compose the accent (e.g. option-n a for ã)

**Comment by smeech** [comment]
Discord discussion starts [here](https://discord.com/channels/884163483409731584/1013916990760554608/1436834122663198941), for future reference.

**Comment by Masqueey** [comment]
Made an issue for dead-keys specifically https://github.com/espanso/espanso/issues/2454. It might be considered a duplicate, although I believe the cause is different (and the solution will be as well) than simply all unicode characters not working. (Which also seems to be fixed since 2.2.2.)

I just checked with 2.3.0 and dead keys still do get captured properly.

**Comment by mthies-unibi** [comment]
As indicated in my comment above, there would be a simple fix using the method `charactersByApplyingModifiers:` to retrieve the just typed character even for a pending dead key, but this would only work for triggers, where the dead key does **not** combine with the next key typed.

For example, <kbd>^</kbd><kbd>2</kbd> could work for a `trigger: "^2"`, but  <kbd>^</kbd><kbd>e</kbd> would invoke `trigger: "^e"` (**not** `trigger: "ê"`) and then too eagerly delete two characters, i.e., the already combined `ê` and the preceding character.

A real fix that works for combining dead keys as well would likely require bigger changes and I have no immediate idea what API on macOS would closely match the Linux/Windows behaviour in this regard.

**Comment by Masqueey** [comment]
I'm not certain I get your solution. For me, your described solution would already work for some of my use-cases. But I don't get how to enable it. Do I need to update line 75 myself by replacing `[event.characters UTF8String]` with `[event charactersByApplyingModifiers: event.modifierFlags]`? Or how would I enable your fix?

**Comment by mthies-unibi** [comment]
Yes, exactly. I experimented with this additional change, while fixing the bug with non-ASCII characters, but since it works only partially (and can delete too many characters in some cases), I did not want to add this change to the official codebase.

If I remember correctly, using a self-built copy of Espanso requires that you ad-hoc sign the newly built Espanso.app (`codesign -s - ...`, see `man codesign`) and to manually add this new copy to the apps with accessibility permissions in the macOS Settings: remove the official Espanso.app from this list with the - button and use the + button to add the new version instead).

This has to be repeated every time the app has been rebuilt. Ad-hoc signing has no concept of application updates that keep their pre-existing permissions. On the other hand, ad-hoc signing works without any specifically obtained developer certificates.
